sway-core/src/ir_generation/function.rs

```diff
@@ -57,6 +57,50 @@ use std::collections::HashMap;
 /// forced, into a value if that is desired. All the temporary values are manipulated with simple
 /// loads and stores, rather than anything more complicated like `mem_copy`s.
 
+// Wrapper around Value to enforce distinction between terminating and non-terminating values.
+enum TerminatorValueWrapper {
+    Terminator(Value),
+    NotTerminator(Value),
+}
+
+impl TerminatorValueWrapper {
+    pub fn new(value: Value, context: &Context) -> Self {
+        if value.is_terminator(context) {
+            Self::Terminator(value)
+        } else {
+            Self::NotTerminator(value)
+        }
+    }
+
+    pub fn is_terminator(&self) -> bool {
+        matches!(self, Self::Terminator(_))
+    }
+
+    pub fn value(&self) -> Value {
+        match self {
+            Self::Terminator(value) | Self::NotTerminator(value) => *value,
+        }
+    }
+}
+
+// If the provided TerminatorValueWrapper is Terminator, then return from the current function immediately.
+// Otherwise extract the embedded Value.
+macro_rules! return_on_termination_or_extract {
+    ($value:expr) => {{
+        let val = $value;
+        if val.is_terminator() {
+            return Ok(val);
+        };
+        val.value()
+    }};
+}
+
+macro_rules! terminator_value_wrap {
+    ($value:expr, $context:expr) => {{
+        TerminatorValueWrapper::new($value, $context)
+    }};
+}
```

Comment:
Since this is just a function call, why does it need to be a macro?

Reply:
It felt more concise to do it this way, but now that I look at it it's not actually that helpful. I'll change it back.